prow/cmd/webhook-server/main.go

```diff
@@ -48,118 +47,158 @@ type ClientInterface interface {
+	if err := optionGroup[0].Validate(o.dryRun); err != nil {
+		return err
+	}
+	if o.expiryInYears < 0 {
+		return fmt.Errorf("invalid expiry years")
+	}
+	return nil
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options	
+	fs.StringVar(&o.projectId, "project-id", "", "Project ID for storing GCP Secrets")
+	fs.IntVar(&o.expiryInYears, "expiry-years", 30, "CA certificate expiry in years")
+	fs.BoolVar(&o.dryRun, "dry-run", true, "Whether to mutate any real-world state")
+	optionGroup := []flagutil.OptionGroup{&o.kubernetes}
+	optionGroup[0].AddFlags(fs)
+	fs.Parse(args)
+	return o
+}
+
 func main() {
 	logrusutil.ComponentInit()
 	logrus.SetLevel(logrus.DebugLevel)
-	p := flag.String("project-id", 
-	"", "Project ID for storing GCP Secrets")
-	e := flag.Int("expiry-years", 30, "CA certificate expiry in years")
-	dns := flagutil.NewStrings("validation-webhook-service", "validation-webhook-service.default", "validation-webhook-service.default.svc")
+	dns := prowflagutil.NewStrings("validation-webhook-service", "validation-webhook-service.default", "validation-webhook-service.default.svc")
 	flag.Var(&dns, "dns", "DNS Names CA-Cert config")
 	flag.Parse()
-	exp := *e
-	projectId := *p
-	if len(projectId) == 0 {
-		logrus.Fatal("project-id flag not supplied")
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	o.dnsNames = dns.Strings()
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
 	}
-	client, err := secretmanager.NewClient(projectId, false)
+	kubeCfg, err := o.kubernetes.InfrastructureClusterConfig(o.dryRun)
 	if err != nil {
-		logrus.WithError(err).Fatal("Unable to create secret manager client")
+		logrus.WithError(err).Fatal("Error getting kubeconfig")
 	}
+	var certFile string
+	var privKeyFile string
 	ctx := context.Background()
-	cert, privKey, err := getGCPSecrets(client, ctx, exp, dns)
-	if err != nil && strings.Contains(err.Error(), gcpSecretError) {
-		logrus.Infof("%v, Will now proceed to create GCP Secret", err)
-		cert, privKey, err = createGCPSecret(client, ctx, exp, dns.Strings())
+	cl, err := ctrlruntimeclient.New(kubeCfg, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not create writer client")
+	}
+	if len(o.projectId) != 0 {
+		client, err := secretmanager.NewClient(o.projectId, false)
 		if err != nil {
-			logrus.WithError(err).Fatal("Unable to create ca certificate")
+			logrus.WithError(err).Fatal("Unable to create secret manager client")
 		}
-	} else if err != nil {
-		logrus.WithError(err).Fatal("Unable to get GCP Secret")
-	}
-	if err = isCertValid(cert); err != nil {
-		logrus.WithError(err).Info("Certificate is not valid, will replace.")
-		cert, privKey, err = updateGCPSecret(client, ctx, exp, dns.Strings())
+		
+		cert, privKey, _, err := getGCPSecrets(client, ctx, o.expiryInYears, dns)
+		var caPem string
+		if err != nil && strings.Contains(err.Error(), gcpSecretError) {
```

Comment:
this is not a reliable way to check whether the secret exists or not, as `getGCPSecrets` could return any other error

Reply:
we checked the error message here and the when the secret isn't there it gives that. wasn't sure of any other way.

Reply:
it would be doable to matching strict error message returns by the GCP secret client, something like `does not exist`. However not very preferable as the error message might change in the future.

Alternatively, we can do:
```
secrets, err := client.List(project)
if err != nil {
  return err // List call should not fail
}
var found bool
for _, secret := range secrets {
  if secret.Name == name {
    found = true
    break
  }
}
```